api/user.go

```diff
@@ -431,114 +430,156 @@ func SendVerifyEmailAndForget(c *Context, userId, userEmail, siteURL string) {
+
+	if len(password) == 0 {
 		c.Err = model.NewLocAppError("login", "api.user.login.blank_pwd.app_error", nil, "")
 		c.Err.StatusCode = http.StatusBadRequest
 		return
 	}
 
-	var user *model.User
+	// bypass all the extra code for handling different login methods when ID is specified
 	if len(props["id"]) != 0 {
-		user = LoginById(c, w, r, props["id"], props["password"], props["token"], props["device_id"])
-	} else if len(props["email"]) != 0 {
-		user = LoginByEmail(c, w, r, props["email"], props["name"], props["password"], props["token"], props["device_id"])
-	} else if len(props["username"]) != 0 {
-		user = LoginByUsername(c, w, r, props["username"], props["name"], props["password"], props["token"], props["device_id"])
+		loginById(c, w, r, props)
+		return
+	}
+
+	ldapAvailable := *utils.Cfg.LdapSettings.Enable && einterfaces.GetLdapInterface() != nil
+
+	c.LogAudit("attempt")
+
+	// get the user
+	var user *model.User
+	if result := <-Srv.Store.User().GetForLogin(
+		loginId,
+		*utils.Cfg.EmailSettings.EnableSignInWithUsername,
+		*utils.Cfg.EmailSettings.EnableSignInWithEmail,
+		ldapAvailable,
+	); result.Err != nil {
+
+		if !ldapAvailable {
+			// failed to find user and no LDAP server to fall back on
+			c.Err = result.Err
+			c.Err.StatusCode = http.StatusBadRequest
+			return
+		}
+
+		// fall back to LDAP server to see if we can find a user
+		if ldapUser, ldapErr := einterfaces.GetLdapInterface().GetUser(loginId); ldapErr != nil {
+			c.Err = ldapErr
+			c.Err.StatusCode = http.StatusBadRequest
+			return
+		} else {
+			user = ldapUser
+		}
 	} else {
-		c.Err = model.NewLocAppError("login", "api.user.login.not_provided.app_error", nil, "")
+		user = result.Data.(*model.User)
+	}
+
+	c.LogAuditWithUserId(user.Id, "attempt")
+
+	// and then authenticate them
+	if user.AuthService == model.USER_AUTH_SERVICE_LDAP {
+		if !ldapAvailable {
+			c.Err = model.NewLocAppError("login", "api.user.login_ldap.not_available.app_error", nil, "")
+			c.Err.StatusCode = http.StatusNotImplemented
+		} else if ldapUser, err := doLdapAuthentication(user.AuthData, password, mfaToken); err != nil {
+			c.Err = err
+			c.Err.StatusCode = http.StatusUnauthorized
+		} else {
+			// slightly redundant to get the user again, but we need to get it from the LDAP server
+			user = ldapUser
+		}
+	} else if user.AuthService != "" {
+		c.Err = model.NewLocAppError("login", "api.user.login.use_auth_service.app_error", map[string]interface{}{"AuthService": user.AuthService}, "")
 		c.Err.StatusCode = http.StatusBadRequest
-		return
+	} else {
+		if err := checkPasswordAndAllCriteria(user, password, mfaToken); err != nil {
+			c.Err = err
+			c.Err.StatusCode = http.StatusUnauthorized
+		}
 	}
 
 	if c.Err != nil {
+		c.LogAuditWithUserId(user.Id, "failure")
 		return
 	}
 
-	if user != nil {
-		user.Sanitize(map[string]bool{})
-	} else {
-		user = &model.User{}
-	}
+	c.LogAuditWithUserId(user.Id, "success")
+
+	Login(c, w, r, user, deviceId)
```

Comment:
I was thinking just DoLogin when I was looking at it before. I think we use a similar naming convention in other places for places that have the same problem although I'm sure we also violate that in a bunch of places. Agree on changing it though since it was also making it a pain to ctrl+f for the login function when I was working on it